src/main/java/org/wso2/carbon/hash/provider/pbkdf2/PBKDF2HashProvider.java

```diff
@@ -0,0 +1,284 @@
+     *
+     * @param iterationCount The iteration count needs to be validated.
+     * @throws HashProviderClientException If the iteration count is negative or equal to zero.
+     */
+    private void validateIterationCount(int iterationCount) throws HashProviderClientException {
+
+        if (iterationCount <= 0) {
+            throw new HashProviderClientException(
+                    ErrorMessage.ERROR_CODE_iNVALID_ITERATION_COUNT.getDescription(),
+                    Constants.PBKDF2_HASH_PROVIDER_ERROR_PREFIX +
+                            ErrorMessage.ERROR_CODE_iNVALID_ITERATION_COUNT.getCode());
+        }
+    }
+
+    /**
+     * This method is responsible for validating the derived key length.
+     *
+     * @param dkLength The derived key length needs to be validated.
+     * @throws HashProviderClientException If the derived key length is negative or equal to zero.
+     */
+    private void validateDerivedKeyLength(int dkLength) throws HashProviderClientException {
+
+        if (dkLength <= 0) {
+            throw new HashProviderClientException(
+                    ErrorMessage.ERROR_CODE_INVALID_DERIVED_KEY_LENGTH.getDescription(),
+                    Constants.PBKDF2_HASH_PROVIDER_ERROR_PREFIX +
+                            ErrorMessage.ERROR_CODE_INVALID_DERIVED_KEY_LENGTH.getCode());
+        }
+    }
+
+    /**
+     * Resolve the iteration count according to the given meta properties.
+     *
+     * @param properties The properties map.
+     * @return The iteration count.
+     */
+    private int resolveIterationCount(Map<String, Object> properties) {
+
+        Object iterationCountObject = properties.get(Constants.ITERATION_COUNT_PROPERTY);
+
+        if (iterationCountObject == null) {
+            return iterationCount;
+        }
+        return (int) properties.get(Constants.ITERATION_COUNT_PROPERTY);
+    }
+
+    /**
+     * Resolve derived key length from the metaProperties.
+     *
+     * @param properties The properties map.
+     * @return The derived key length.
+     */
+    private int resolveDerivedKeyLength(Map<String, Object> properties) {
+
+        Object dkLengthObject = properties.get(Constants.DERIVED_KEY_LENGTH_PROPERTY);
+
+        if (dkLengthObject == null) {
+            return dkLength;
+        }
+        return (int) properties.get(Constants.DERIVED_KEY_LENGTH_PROPERTY);
+    }
+
+    /**
+     * Resolve pseudo random function from the metaProperties.
+     *
+     * @param properties The properties map.
+     * @return The pseudo random function.
+     */
+    private String resolvePseudoRandomFunction(Map<String, Object> properties) {
+
+        Object pseudoRandomFunctionObject = properties.get(Constants.PSEUDO_RANDOM_FUNCTION_PROPERTY);
+
+        if (pseudoRandomFunctionObject == null) {
+            return pseudoRandomFunction;
+        }
+        return (String) properties.get(Constants.PSEUDO_RANDOM_FUNCTION_PROPERTY);
```

Comment:
Don't we need to handle if this iterationCountObject cannot cast to an String?

Reply:
when casting can we use `pseudoRandomFunctionObject` since we already have that?